src/compiler/checker.ts

```diff
@@ -16097,7 +16092,8 @@ namespace ts {
                 const constraint = getConstraintOfTypeParameter(inference.typeParameter);
                 if (constraint) {
                     const instantiatedConstraint = instantiateType(constraint, context.nonFixingMapper);
-                    if (!inferredType || !context.compareTypes(inferredType, getTypeWithThisArgument(instantiatedConstraint, inferredType))) {
+                    const isCompletionContext = contextFlags && (contextFlags & ContextFlags.Completion);
+                    if (!inferredType || !context.compareTypes(inferredType, getTypeWithThisArgument(instantiatedConstraint, inferredType)) || isCompletionContext) {

```

Comment:
Nit: put `isCompletionContext` before `!context.compareTypes(...)` for more efficient short-circuiting for completions

Reply:
Good point! Just fixed it.